File: sdk/python/src/tinyverse/api/groups.py
from __future__ import annotations

from typing import Any, Optional
from urllib.parse import quote as url_encode

from ..http import HttpClient


class GroupsApi:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def list(self, params: Optional[dict[str, Any]] = None) -> Any:
        return await self._http.get("/directory/groups", params)

    async def get(self, group_id: str) -> Any:
        return await self._http.get(f"/directory/groups/{url_encode(group_id, safe='')}")

    async def create(self, request: dict[str, Any]) -> Any:
        return await self._http.post("/directory/groups", request)

    async def members(self, group_id: str) -> Any:
        return await self._http.get(
            f"/directory/groups/{url_encode(group_id, safe='')}/members",
        )

    async def join(self, group_id: str) -> Any:
        return await self._http.post(
            f"/directory/groups/{url_encode(group_id, safe='')}/join",
        )

File: sdk/rust/src/api/groups.rs
use std::sync::Arc;

use serde::{Deserialize, Serialize};

use crate::error::Result;
use crate::http::{HttpClient, to_body, to_query};
use crate::types::{GroupCreateRequest, GroupMember, GroupMetadata, GroupQueryParams};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupListResponse {
    pub groups: Vec<GroupMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupMemberListResponse {
    pub members: Vec<GroupMember>,
}

pub struct GroupsApi {
    http: Arc<HttpClient>,
}

impl GroupsApi {
    pub fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    pub async fn list(&self, params: Option<&GroupQueryParams>) -> Result<GroupListResponse> {
        let query = params.map(to_query).transpose()?;
        self.http.get("/directory/groups", query.as_ref()).await
    }

    pub async fn get(&self, group_id: &str) -> Result<GroupMetadata> {
        let path = format!("/directory/groups/{}", urlencoding::encode(group_id));
        self.http.get(&path, None).await
    }

    pub async fn create(&self, request: &GroupCreateRequest) -> Result<GroupMetadata> {
        self.http
            .post("/directory/groups", Some(&to_body(request)?))
            .await
    }

    pub async fn members(&self, group_id: &str) -> Result<GroupMemberListResponse> {
        let path = format!(
            "/directory/groups/{}/members",
            urlencoding::encode(group_id)
        );
        self.http.get(&path, None).await
    }

    pub async fn join(&self, group_id: &str) -> Result<GroupMember> {
        let path = format!(
            "/directory/groups/{}/join",
            urlencoding::encode(group_id)
        );
        self.http.post(&path, None).await
    }
}
